### ISSUE TITLE
Need some internals documentation for new developers

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#6814]
As a new developer, I find it take a good bit of research (reading code/comments, sometimes using the debugger) to make headway in understanding the code.  The few short documents that exist in the source tree cover almost nothing.

It would help a lot to have a document that gives an overview of the system as a whole:
- Identifies the major parts of the code base such as: vernacular/parsing, the low-level proof validation engine, tactics, Ltac stuff, the state manager (stm), coqide, how to do plugins etc. with a basic description, their important data structures and interfaces, etc. and a bit about how they relate to each other.  Diagrams would be great provided there are good tools for that and that the diagrams can be edited in the future
- Describes how to do basic development tasks such as adding a new tactic, command or plugin
- Where to start in understanding the system or undertaking some change
- Maybe notes identifying parts of the system that are crufty or poorly understood

Parsing, in particular, is baffling.  I understand a bit about what camlp5 does in general, but "Qed" and "subst" are associated with actions is unclear.  (I've not had to add a command or tactic yet.)  Also, it's not so helpful that the camlp5 grammar is apparently not in the Coq source tree.

Also, what's the deal with the .v files that contain binary data?  That makes it pretty much impossible to understand what that code does.

I think creating overview documentation will be an iterative process--as newcomers or others ask questions, the document should be extended.

**Comment by jfehrle** [comment]
It would also be immensely helpful if there was a routine that could print the value of any variable of any type.  If it's possible to serialize any value, it should be possible to print any value.

**Comment by SkySkimmer** [comment]
>Also, what's the deal with the .v files that contain binary data? That makes it pretty much impossible to understand what that code does.

Do you mean .ml files (otherwise I don't know which files you're talking about)? The camlp5 .ml4 and .mlp files are processed producing an ocaml binary ast which is put in .ml files (so foo.ml4 produces foo.ml) during compilation. You can get regular textual ocaml code by compiling with READABLE_ML4=1 (it's generated code so it's not that readable though).
(This is already explained in the dev/doc/build system files)

**Comment by Zimmi48** [comment]
There is already a beginning of developer documentation at https://github.com/coq/coq/tree/master/dev/doc. Is this what you mean by "the few short documents in the source tree"? These documents are quite heterogeneous, not all of them are very fresh, and it's hard to browse because there is no real index.

I suppose the job could start by updating these documents, converting them to Markdown, adding a README in this folder with links to the various available pieces.

The overview of the system is at https://github.com/coq/coq/blob/master/dev/doc/coq-src-description.txt but it is very incomplete AFAICS (the STM is missing for instance). It would be good to have links from this file to relevant pieces of documentation (in other files or in the source code itself).

There's also the issue of the current code structure evolving very fast these days, notably on @ejgallego's impulse. So such descriptions are bound to go outdated and should be updated often (preferably at the same time as the structure of the code base itself).

**Comment by ejgallego** [comment]
Indeed Coq is challenging project to get into it, even with good documentation, the size of the codebase is no toy. We are trying to improve documentation indeed, but maybe the biggest problem as of today is that all the knowledge about the codebase is _scattered_ among mailing lists, GH, SO, and in the codebase itself in different formats and files [standalone files, .ml files, .tex files...].

For now I can give you two suggestions:
- serialization / printing can be achieved either by using `Drop` + the debug printers, or in a different style, using `SerAPI`, however while SerAPI is less tailored to low-level development, it may be useful for people living in the upper layers.
- Ask! We have both gitter and mailing lists. Indeed, the most effective way to transmit knowledge about the system is face-to-face. So you should try to attend developer meetings, and ask questions in the proper channels. No amount of documentation will be complete w.r. all the many intricacies of Coq.

**Comment by maximedenes** [comment]
> serialization / printing can be achieved either by using Drop + the debug printers, or in a different style, using SerAPI, however while SerAPI is less tailored to low-level development, it may be useful for people living in the upper layers.

Also with `ocamldebug`.

**Comment by Zimmi48** [comment]
> So you should try to attend developer meetings, and ask questions in the proper channels.

We should try to organize a Coq Implementors Workshop in the US like we do in France every year. I think there would be sufficiently many people interested over there.

**Comment by jfehrle** [comment]
(responding to several messages)

Complementing top-down documentation, maybe we could also put more bottom-up documentation in the source code comments.  Java has a Javadoc tool that extracts this information and turns them into web pages (one per source file IIRC) with a full index (output examples: https://docs.oracle.com/javase/7/docs/api/java/lang/package-summary.html and https://docs.oracle.com/javase/7/docs/api/java/lang/String.html).  Does OCaml have something similar?  (How is https://caml.inria.fr/pub/docs/manual-ocaml/libref/ generated?)

Printing with ocamldebug is helpful, though there's significant overhead (sometimes prohibitive) to restart the debugger after each code change, then set and step through appropriate breakpoints.  Especially with have a limited understanding of the flow of control.  I've not yet gotten to a point where Drop is helpful.  So having a general print routine would be very helpful.  I'm surprised no one has created this yet.  SerAPI may be useful to me later; for right now it doesn't look useful because a) it is specific to Coq b) I really want to see all the OCaml-level details, which I'm still learning.  Any abstraction in the output at this point is confusing.

OCaml itself needs better reference documentation.  There is quite a bit of information in the documents, but it's hard to find things.  The doc needs to be more complete, in places written more clearly and better indexed.  And more examples.  I'm thinking about working on improving Chapter 7/8 of the OCaml documentation.

Understanding a new code base is indeed challenging.  Tangent: We spend too much time identifying basic properties, such as what routines some bit of code calls, what are its side effects and assumptions, etc.  Someday we'll have tools to help more with that, and I expect they'll need to use proof techniques to infer some properties.

Maybe worth mentioning gitter here: https://coq.inria.fr/community.  But I found the reference to it in CONTRIBUTING.md.

Yes, I meant binary in .ml files, not .v files.

**Comment by Zimmi48** [comment]
There is ocamldoc. We are using it already a little (comments starting with `(**`). There is a `mli-doc` target in the Makefile. That's typically something else that should be mentioned in a README in `dev/doc`.

> Maybe worth mentioning gitter here: https://coq.inria.fr/community. But I found the reference to it in CONTRIBUTING.md.

Good point. I suggest you open a PR at http://github.com/coq/www to add the mention (the community page can be found in the `pages/` directory).

**Comment by jfehrle** [comment]
> I suggest you open a PR at http://github.com/coq/www to add the mention (the community page can be found in the pages/ directory).
Done.  I see 2 other rather old PRs for that repository, BTW.

**Comment by Zimmi48** [comment]
> I see 2 other rather old PRs for that repository, BTW.

Yes, this repository has been badly managed but it's getting better. You should have seen the number of old open PRs a few months ago!
